# Crime Analysis Report

In [25]:
import pandas as pd

In [26]:
df = pd.read_csv("./dataset/DM_Dataset-T12.csv")
df.head(n=10)

STATE        DISTRICT  Year  Murder  Infanticide  Rape  Foeticide  \
0  ANDHRA PRADESH        ADILABAD  2001     0.0          NaN   0.0        0.0   
1  ANDHRA PRADESH       ANANTAPUR  2001    19.0          NaN  12.0        0.0   
2  ANDHRA PRADESH        CHITTOOR  2001     0.0          NaN   0.0        0.0   
3  ANDHRA PRADESH        CUDDAPAH  2001     0.0          NaN   0.0        0.0   
4  ANDHRA PRADESH   EAST GODAVARI  2001     0.0          NaN   0.0        0.0   
5  ANDHRA PRADESH   GUNTAKAL RLY.  2001     0.0          NaN   0.0        0.0   
6  ANDHRA PRADESH          GUNTUR  2001     0.0          NaN   0.0        0.0   
7  ANDHRA PRADESH  HYDERABAD CITY  2001     2.0          NaN   0.0        0.0   
8  ANDHRA PRADESH      KARIMNAGAR  2001     1.0          NaN   4.0        0.0   
9  ANDHRA PRADESH         KHAMMAM  2001     0.0          NaN   0.0        0.0   

   Kidnapping and Abduction  Abetment of suicide  Exposure and abandonment  \
0                       0.0                  0.0                       0.0   
1                      29.0                  6.0                       0.0   
2                       0.0                  0.0                       0.0   
3                       0.0                  0.0                       0.0   
4                       0.0                  0.0                       0.0   
5                       0.0                  0.0                       0.0   
6                       0.0                  0.0                       0.0   
7                       0.0                  0.0                      19.0   
8                       0.0                  0.0                       3.0   
9                       0.0                  0.0                       0.0   

   Procuration of minor girls  Selling of girls for prostitution  \
0                         0.0                                0.0   
1                         0.0                                0.0   
2                         0.0                                0.0   
3                         0.0                                0.0   
4                         0.0                                0.0   
5                         0.0                                0.0   
6                         0.0                                0.0   
7                         2.0                                0.0   
8                         0.0                                0.0   
9                         0.0                                0.0   

   Prohibition of child marriage act  Other Crimes  Total  \
0                                0.0           0.0      0   
1                                0.0           0.0     66   
2                                0.0           0.0      0   
3                                0.0           0.0      0   
4                                0.0           0.0      0   
5                                0.0           0.0      0   
6                                0.0           0.0      0   
7                                2.0           0.0     25   
8                                0.0           1.0      9   
9                                0.0           0.0      0   

   Literacy Rate- Total  Literacy Rate - Rural  Literacy Rate - Urban  
0                   NaN                    NaN                    NaN  
1                   NaN                    NaN                    NaN  
2                   NaN                    NaN                    NaN  
3                   NaN                    NaN                    NaN  
4                   NaN                    NaN                    NaN  
5                   NaN                    NaN                    NaN  
6                   NaN                    NaN                    NaN  
7                   NaN                    NaN                    NaN  
8                   NaN                    NaN                    NaN  
9                   NaN                    NaN                    NaN

## Dataset Analysis

The dataset consists of data corresponding to crimes recorded from the year 2001-2015. The dataset is distributed district-wise for each state of India at the time of collection of data. The original dataset has been appended with the last 3 columns which contain the literacy rate present in the state based on 2001 and 2011 census.

In [27]:
df.shape

(11527, 18)

There are a total of 11,527 entries corresponding to the 15 years of data. As seen above, there are 18 columns (or attributes) present in the data which represents data corresponding to various crimes commited like murder, rape, foeticide etc. Most of the attributes are numerical in nature.

In [28]:
df['STATE'].unique()

array(['ANDHRA PRADESH', 'ARUNACHAL PRADESH', 'ASSAM', 'BIHAR',
       'CHHATTISGARH', 'GOA', 'GUJARAT', 'HARYANA', 'HIMACHAL PRADESH',
       'JAMMU & KASHMIR', 'JHARKHAND', 'KARNATAKA', 'KERALA',
       'MADHYA PRADESH', 'MAHARASHTRA', 'MANIPUR', 'MEGHALAYA', 'MIZORAM',
       'NAGALAND', 'ODISHA', 'PUNJAB', 'RAJASTHAN', 'SIKKIM',
       'TAMIL NADU', 'TRIPURA', 'UTTAR PRADESH', 'UTTARAKHAND',
       'WEST BENGAL', 'A & N ISLANDS', 'CHANDIGARH', 'D & N HAVELI',
       'DAMAN & DIU', 'DELHI', 'LAKSHADWEEP', 'PUDUCHERRY',
       'Andhra Pradesh', 'Arunachal Pradesh', 'Assam', 'Bihar',
       'Chhattisgarh', 'Goa', 'Gujarat', 'Haryana', 'Himachal Pradesh',
       'Jammu & Kashmir', 'Jharkhand', 'Karnataka', 'Kerala',
       'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram',
       'Nagaland', 'Odisha', 'Punjab', 'Rajasthan', 'Sikkim',
       'Tamil Nadu', 'Tripura', 'Uttar Pradesh', 'Uttarakhand',
       'West Bengal', 'A&N Islands', 'Chandigarh', 'D&N Haveli',
       '

There are multiple names corresponding to a single state. Also, some names are capitalized. These are preprocessed as follows:

In [29]:
for i in range (len(df)):
    df.loc[i, 'STATE'] = df.loc[i, 'STATE'].lower()
    df.loc[i, 'DISTRICT'] = df.loc[i, 'DISTRICT'].lower()
    if df.loc[i, 'STATE'] == 'a & n islands' : df.loc[i, 'STATE'] = 'a&n islands'
    if df.loc[i, 'STATE'] == 'delhi ut' : df.loc[i, 'STATE'] = 'delhi'
    if df.loc[i, 'STATE'] == 'd & n haveli' : df.loc[i, 'STATE'] = 'd&n haveli'
    if df.loc[i, 'DISTRICT'] == 'total district(s)' : df.loc[i, 'DISTRICT'] = 'total'
    if df.loc[i, 'DISTRICT'] == 'zdelhi ut total' : df.loc[i, 'DISTRICT'] = 'total'
        
df['STATE'].unique()

array(['andhra pradesh', 'arunachal pradesh', 'assam', 'bihar',
       'chhattisgarh', 'goa', 'gujarat', 'haryana', 'himachal pradesh',
       'jammu & kashmir', 'jharkhand', 'karnataka', 'kerala',
       'madhya pradesh', 'maharashtra', 'manipur', 'meghalaya', 'mizoram',
       'nagaland', 'odisha', 'punjab', 'rajasthan', 'sikkim',
       'tamil nadu', 'tripura', 'uttar pradesh', 'uttarakhand',
       'west bengal', 'a&n islands', 'chandigarh', 'd&n haveli',
       'daman & diu', 'delhi', 'lakshadweep', 'puducherry', 'telangana'],
      dtype=object)

In [30]:
df[['Infanticide', 'Literacy Rate- Total', 'Literacy Rate - Rural', 'Literacy Rate - Urban']].describe()

Infanticide  Literacy Rate- Total  Literacy Rate - Rural  \
count  2512.000000             55.000000              55.000000   
mean      0.234076             71.752727              67.330909   
std       1.368759             10.122671              10.805398   
min       0.000000             47.000000              43.900000   
25%       0.000000             66.250000              60.150000   
50%       0.000000             70.500000              66.800000   
75%       0.000000             77.450000              73.450000   
max      33.000000             94.000000              93.000000   

       Literacy Rate - Urban  
count              55.000000  
mean               83.752727  
std                 6.031825  
min                69.800000  
25%                79.550000  
50%                83.900000  
75%                87.750000  
max                97.600000

## Missing Values
The columns Literacy Rate - Total, Literacy Rate - Urban and Literacy Race - Rural have only 55 entries. Thus, their data is very sparse and we opt to drop these 3 columns. We can use their values for useful insights during visualization plots later.

Also, the column Total is the sum of values of different crimes in each column and thus is redundant. We are also going to drop it.

The column Infanticide contains data only from the year 2013-2015 and has NA for years 2001-2012. Thus we are dropping it too. We may use it for year specific analysis in years from 2013-2015.

In [31]:
df = df.drop(['Infanticide', 'Literacy Rate- Total', 'Literacy Rate - Rural', 'Literacy Rate - Urban'], axis=1)

In [32]:
df.columns

Index(['STATE', 'DISTRICT', 'Year', 'Murder', 'Rape', 'Foeticide',
       'Kidnapping and Abduction', 'Abetment of suicide',
       'Exposure and abandonment', 'Procuration of minor girls',
       'Selling of girls for prostitution',
       'Prohibition of child marriage act', 'Other Crimes', 'Total'],
      dtype='object')

Upon closer observation on the dataset, we see that for each state there is a row corresponding to the total amount of crimes in each column. We are also going to remove this entry for each state and for each year. Also the rows containing atleast one NA values are removed.

In [33]:
df = df[df['DISTRICT'] != 'total']
df.dropna(inplace=True)
df.reset_index(drop=True, inplace=True)

In [34]:
df.shape

(10995, 14)

## Aggregation
We have thus a total of 10995 entries. However, the dataset is quite sparse with large amount of zeroes. Thus, we perform aggregation and combine multiple entries within each state into a single entry so as to reduce the sparsity of the data. 

However, this combination is not possible on a district wise manner since for combining districts we need to find its neighbouring districts which is not possible based on the given data. Thus, we analyze data and produce results at a state level, disregarding the district in which the crime had occurred.

In [35]:
data_frame = []
i = 0
while i < len(df):
    state = df.loc[i, 'STATE']
    year = df.loc[i, 'Year']
    count = 0
    for j in range(i, i+4):
        if j < len(df) and df.loc[j, 'STATE'] == state and df.loc[j, 'Year'] == year:
            count = count + 1
    temp = df.iloc[i: i+count].drop(['STATE', 'DISTRICT', 'Year'], axis=1).sum().to_list()
    temp.insert(0, state)
    temp.insert(1, int(year))
    data_frame.append(temp)
    if count == 0:
        i = i + 1
    else:
        i = i + count
    
df_columns = list(df.columns[3:14])
df_columns.insert(0, 'State')
df_columns.insert(1, 'Year')
df_aggregated = pd.DataFrame(data_frame, columns=df_columns)
df_aggregated.head(n=50)

State  Year  Murder  Rape  Foeticide  \
0      andhra pradesh  2001    19.0  12.0        0.0   
1      andhra pradesh  2001     2.0   0.0        0.0   
2      andhra pradesh  2001     2.0  15.0        0.0   
3      andhra pradesh  2001     1.0  34.0        0.0   
4      andhra pradesh  2001     5.0   9.0        0.0   
5      andhra pradesh  2001     5.0   8.0        0.0   
6      andhra pradesh  2001     1.0   6.0        0.0   
7   arunachal pradesh  2001     0.0   0.0        0.0   
8   arunachal pradesh  2001     0.0   0.0        0.0   
9   arunachal pradesh  2001     0.0   0.0        0.0   
10  arunachal pradesh  2001     0.0   0.0        0.0   
11              assam  2001     0.0   0.0        0.0   
12              assam  2001     0.0   0.0        0.0   
13              assam  2001     0.0   0.0        0.0   
14              assam  2001     0.0   0.0        0.0   
15              assam  2001     0.0   0.0        0.0   
16              assam  2001     0.0   0.0        0.0   
17              assam  2001     0.0   0.0        0.0   
18              bihar  2001     0.0   4.0        0.0   
19              bihar  2001     0.0   3.0        0.0   
20              bihar  2001     0.0   0.0        0.0   
21              bihar  2001     0.0   0.0        0.0   
22              bihar  2001     0.0   0.0        0.0   
23              bihar  2001     1.0   1.0        0.0   
24              bihar  2001     0.0   0.0        0.0   
25              bihar  2001     0.0   3.0        0.0   
26              bihar  2001     1.0   5.0        0.0   
27              bihar  2001     0.0   0.0        0.0   
28              bihar  2001     0.0   0.0        0.0   
29       chhattisgarh  2001     0.0  19.0        2.0   
30       chhattisgarh  2001     2.0   6.0        0.0   
31       chhattisgarh  2001     1.0  10.0        1.0   
32       chhattisgarh  2001     3.0  38.0        0.0   
33       chhattisgarh  2001     8.0  77.0        2.0   
34                goa  2001     1.0  10.0        0.0   
35            gujarat  2001     8.0   2.0        0.0   
36            gujarat  2001     8.0   6.0        0.0   
37            gujarat  2001    17.0   5.0        2.0   
38            gujarat  2001     4.0   0.0        0.0   
39            gujarat  2001     5.0   2.0        0.0   
40            gujarat  2001    30.0  12.0        2.0   
41            gujarat  2001     6.0  12.0        0.0   
42            gujarat  2001     0.0   0.0        0.0   
43            haryana  2001     7.0  26.0        3.0   
44            haryana  2001     5.0   6.0        0.0   
45            haryana  2001     9.0  42.0        0.0   
46            haryana  2001    10.0  33.0        0.0   
47            haryana  2001     0.0   1.0        0.0   
48   himachal pradesh  2001     1.0   8.0        0.0   
49   himachal pradesh  2001     2.0  12.0        0.0   

    Kidnapping and Abduction  Abetment of suicide  Exposure and abandonment  \
0                       29.0                  6.0                       0.0   
1                        0.0                  0.0                      19.0   
2                        2.0                  0.0                       3.0   
3                       16.0                  0.0                       0.0   
4                        6.0                  1.0                       0.0   
5                        4.0                  0.0                       0.0   
6                        0.0                  0.0                       0.0   
7                        0.0                  0.0                       0.0   
8                        0.0                  0.0                       0.0   
9                        0.0                  0.0                       0.0   
10                       0.0                  0.0                       0.0   
11                       0.0                  0.0                       0.0   
12                       4.0                  0.0                       0.0   
13                       3.0                  0.0    

In [36]:
df_aggregated.shape

(2971, 13)

Thus we see after aggregation, the total number of entries has been reduced from 10995 to 2971. The overall statistics corresponding to these are as follows:

In [37]:
df_aggregated.iloc[:,2:].describe()

Murder         Rape    Foeticide  Kidnapping and Abduction  \
count  2971.000000  2971.000000  2971.000000               2971.000000   
mean      7.548300    31.808145     0.570178                 66.185796   
std       9.147645    44.240362     1.728560                159.718533   
min       0.000000     0.000000     0.000000                  0.000000   
25%       1.000000     2.000000     0.000000                  3.000000   
50%       5.000000    16.000000     0.000000                 19.000000   
75%      11.000000    43.000000     0.000000                 61.000000   
max      76.000000   531.000000    21.000000               2705.000000   

       Abetment of suicide  Exposure and abandonment  \
count          2971.000000               2971.000000   
mean              0.296197                  4.208684   
std               1.620299                  8.267673   
min               0.000000                  0.000000   
25%               0.000000                  0.000000   
50%               0.000000                  0.000000   
75%               0.000000                  5.000000   
max              67.000000                 75.000000   

       Procuration of minor girls  Selling of girls for prostitution  \
count                 2971.000000                        2971.000000   
mean                     3.504881                           0.357792   
std                     19.498896                           2.760634   
min                      0.000000                           0.000000   
25%                      0.000000                           0.000000   
50%                      0.000000                           0.000000   
75%                      0.000000                           0.000000   
max                    433.000000                          69.000000   

       Prohibition of child marriage act  Other Crimes        Total  
count                        2971.000000   2971.000000  2971.000000  
mean                            0.644901     33.102659   165.475934  
std                             1.823008     70.058835   270.765567  
min                             0.000000      0.000000     0.000000  
25%                             0.000000      0.000000    15.000000  
50%                             0.000000      5.000000    79.000000  
75%                             1.000000     29.000000   205.000000  
max                            29.000000    729.000000  3770.000000

## Binarization
In order to apply data mining algorithms like associate rule mining, we need to convert our continuous valued data into binary data through the process of binarization.

The threshold value for an attribute j in an entry i is given by:
Let total crimes in entry i for all attributes be p
Let total crimes of all entries for all attributes in the state (for particular year) be z
Let total crimes of all entries for attribute j be w
Then threshold = (p/z)*w

In [38]:
i = 0
while i < len(df_aggregated):
    state = df_aggregated.loc[i, 'State']
    year = df_aggregated.loc[i, 'Year']
    j = i
    count = 0
    
    while j < len(df_aggregated) and df_aggregated.loc[j, 'State'] == state and df_aggregated.loc[j, 'Year'] == year:
        count = count + 1
        j = j + 1
    
    temp = df_aggregated.iloc[i:i+count].drop(['State', 'Year'], axis=1).sum().to_list()
    j = i
    while j < len(df_aggregated) and df_aggregated.loc[j, 'State'] == state and df_aggregated.loc[j, 'Year'] == year:
        for k in range (df_aggregated.shape[1] - 3):
            if df_aggregated.iloc[j][k+2] > ((df_aggregated.iloc[j][12] / temp[10]) * temp[k]):
                df_aggregated.iat[j, k+2] = 1
            else:
                df_aggregated.iat[j, k+2] = 0
        j = j + 1
    i = i + count
    
df_aggregated = df_aggregated.drop(['Total'], axis=1)

/home/daksh/.local/share/virtualenvs/crime-analysis-report-ZCftCXec/lib/python3.7/site-packages/ipykernel_launcher.py:16: RuntimeWarning: invalid value encountered in double_scalars
  app.launch_new_instance()


Thus, now each entry can be considered as a transaction and can be used for association analysis. We further remove all transactions that contain no items, that is all of their entries are zero.

In [43]:
df_binarized = df_aggregated.drop(['State', 'Year'], axis=1)
for i in range (len(df_binarized)):
    var = False
    for j in range(len(list(df_binarized.iloc[i]))):
        if df_binarized.iloc[i][j] != 0:
            var = True
    if var == False:
        df_aggregated = df_aggregated.drop(df.index[[int(i)]])

In [44]:
df_aggregated.head(n=30)

State  Year  Murder  Rape  Foeticide  Kidnapping and Abduction  \
0   andhra pradesh  2001     1.0   0.0        0.0                       1.0   
1   andhra pradesh  2001     0.0   0.0        0.0                       0.0   
2   andhra pradesh  2001     0.0   1.0        0.0                       0.0   
3   andhra pradesh  2001     0.0   1.0        0.0                       1.0   
4   andhra pradesh  2001     0.0   0.0        0.0                       0.0   
5   andhra pradesh  2001     1.0   1.0        0.0                       0.0   
6   andhra pradesh  2001     0.0   1.0        0.0                       0.0   
18           bihar  2001     0.0   1.0        0.0                       1.0   
19           bihar  2001     0.0   0.0        0.0                       0.0   
23           bihar  2001     1.0   1.0        0.0                       0.0   
24           bihar  2001     0.0   0.0        0.0                       0.0   
25           bihar  2001     0.0   0.0        0.0                       1.0   
26           bihar  2001     1.0   1.0        0.0                       0.0   
27           bihar  2001     0.0   0.0        0.0                       1.0   
29    chhattisgarh  2001     0.0   1.0        1.0                       1.0   
30    chhattisgarh  2001     0.0   0.0        0.0                       0.0   
31    chhattisgarh  2001     1.0   1.0        1.0                       1.0   
32    chhattisgarh  2001     1.0   1.0        0.0                       1.0   
33    chhattisgarh  2001     1.0   1.0        1.0                       1.0   
35         gujarat  2001     0.0   0.0        0.0                       0.0   
36         gujarat  2001     0.0   0.0        0.0                       0.0   
37         gujarat  2001     1.0   1.0        1.0                       0.0   
38         gujarat  2001     0.0   0.0        0.0                       1.0   
39         gujarat  2001     1.0   0.0        0.0                       1.0   
40         gujarat  2001     1.0   1.0        1.0                       0.0   
41         gujarat  2001     0.0   1.0        0.0                       1.0   
43         haryana  2001     0.0   0.0        1.0                       1.0   
44         haryana  2001     1.0   0.0        0.0                       1.0   
45         haryana  2001     1.0   1.0        0.0                       0.0   
46         haryana  2001     1.0   1.0        0.0                       0.0   

    Abetment of suicide  Exposure and abandonment  Procuration of minor girls  \
0                   1.0                       0.0                         0.0   
1                   0.0                       1.0                         1.0   
2                   0.0                       1.0                         1.0   
3                   0.0                       0.0                         1.0   
4                   0.0                       0.0                         0.0   
5                   0.0                       0.0                         0.0   
6                   0.0                       0.0                         1.0   
18                  0.0                       0.0                         0.0   
19                  0.0                       0.0                         0.0   
23                  0.0                       0.0                         0.0   
24                  0.0                       0.0                         0.0   
25                  0.0                       1.0                         0.0   
26                  0.0                       0.0                         1.0   
27                  0.0                       0.0                         0.0   
29                  1.0                       1.0                         0.0   
30                  0.0                       0.0                         0.0   
31                  0.0                       1.0                         0.0   
32                  0.0                       0.0                         0.0   
33                  0.0 

In [45]:
df_aggregated.tail(n=30)

State  Year  Murder  Rape  Foeticide  Kidnapping and Abduction  \
2934  uttar pradesh  2015     1.0   1.0        0.0                       0.0   
2935  uttar pradesh  2015     1.0   1.0        0.0                       0.0   
2936  uttar pradesh  2015     1.0   0.0        1.0                       1.0   
2937  uttar pradesh  2015     0.0   1.0        0.0                       1.0   
2938  uttar pradesh  2015     1.0   0.0        1.0                       1.0   
2940  uttar pradesh  2015     0.0   0.0        1.0                       1.0   
2941  uttar pradesh  2015     1.0   0.0        0.0                       0.0   
2942  uttar pradesh  2015     1.0   0.0        0.0                       0.0   
2943  uttar pradesh  2015     0.0   1.0        0.0                       1.0   
2944  uttar pradesh  2015     1.0   1.0        0.0                       0.0   
2945  uttar pradesh  2015     1.0   0.0        0.0                       0.0   
2946  uttar pradesh  2015     0.0   0.0        0.0                       0.0   
2947  uttar pradesh  2015     0.0   1.0        0.0                       1.0   
2948    uttarakhand  2015     1.0   1.0        0.0                       1.0   
2949    uttarakhand  2015     1.0   0.0        0.0                       0.0   
2950    uttarakhand  2015     0.0   1.0        0.0                       1.0   
2951    uttarakhand  2015     0.0   0.0        0.0                       1.0   
2952    west bengal  2015     1.0   0.0        0.0                       1.0   
2953    west bengal  2015     1.0   0.0        0.0                       1.0   
2954    west bengal  2015     0.0   0.0        0.0                       0.0   
2955    west bengal  2015     1.0   0.0        0.0                       1.0   
2956    west bengal  2015     0.0   0.0        0.0                       0.0   
2957    west bengal  2015     1.0   0.0        0.0                       0.0   
2958    west bengal  2015     0.0   0.0        0.0                       0.0   
2959    west bengal  2015     0.0   0.0        0.0                       0.0   
2964          delhi  2015     0.0   0.0        0.0                       0.0   
2965          delhi  2015     1.0   0.0        1.0                       1.0   
2966          delhi  2015     0.0   0.0        0.0                       1.0   
2967          delhi  2015     1.0   1.0        0.0                       0.0   
2968          delhi  2015     1.0   1.0        1.0                       1.0   

      Abetment of suicide  Exposure and abandonment  \
2934                  0.0                       0.0   
2935                  0.0                       0.0   
2936                  1.0                       0.0   
2937                  0.0                       1.0   
2938                  0.0                       0.0   
2940                  0.0                       0.0   
2941                  0.0                       0.0   
2942                  0.0                       0.0   
2943                  0.0                       0.0   
2944                  0.0                       0.0   
2945                  0.0                       0.0   
2946                  1.0                       0.0   
2947                  0.0                       0.0   
2948                  0.0                       0.0   
2949                  0.0                       0.0   
2950                  0.0                       0.0   
2951                  0.0                       0.0   
2952                  0.0                       0.0   
2953                  0.0                       0.0   
2954                  1.0                       0.0   
2955                  0.0                       0.0   
2956                  1.0                       0.0   
2957                  0.0                       0.0   
2958                  0.0                       0.0   
2959                  0.0                       0.0   
2964                  0.0                       0.0   
2965                  0.0                       

In [46]:
df_aggregated.shape

(2596, 12)